File: 0001-two-sum/0001-two-sum.cpp
// TC: O(N), SC: O(N)

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mp;
        vector<int> res;
        for(int i = 0; i < nums.size(); i++) {
            if(mp.find(target - nums[i]) != mp.end()) { // hashing the indices reduces calculations
                res.push_back(mp[target - nums[i]]);
                res.push_back(i);
            }
            mp[nums[i]] = i;
        }
        return res;
    }
};

File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> mp = new HashMap<>();
        int[] res = new int[2];

        for(int i = 0; i < nums.length; i++) {
            if(mp.containsKey(target - nums[i])) {
                res[0] = mp.get(target - nums[i]);
                res[1] = i;
            }
            mp.put(nums[i], i);
        }

        return res;
    }
}